fix(gatsby): ignore crawlers when prefetching

me4502 opened the pull request:
## Description

Adding prefetch links to all linked pages is great for the end user, but for bots this is often useless and a waste of bandwidth. For SEO purposes, this can actually be actively harmful, especially on sites such as blocks that might have text-heavy page-data.json files. As search engine crawlers such as Googlebot don't make use of these files, anything it downloads will be wasted bandwidth. As many SEO crawlers have daily allowance limits, this means on larger sites you're going to hit issues where the bandwidth allowance is used up by prefetched files that go entirely unused.

In my testcase of https://clipchamp.com/en/, 37.8kB of the total 65.3kB of bandwidth used to load the page were from prefetched files. This is 57%, well over half. This means that Googlebot will crawl on average, less than half the pages it could crawl had these prefetch calls not been added. I imagine this would be fairly similar for other large sites that include many internal links and content-heavy pages. In our findings, nearly all hits to page-data files by Googlebot were not accompanied by a request to the page it was associated with.

By preventing prefetch links from being added on bots, these issues no longer exist.

### Documentation

Likely N/A

<!--
  Where is this feature or API documented?

  - If docs exist:
    - Update any references, if relevant. This includes Guides and Gatsby Internals docs.
  - If no docs exist:
    - Create a stub for documentation including bullet points for how to use the feature, code snippets (including from happy path tests), etc.
-->

me4502 commented:
Unless these test failures were caused by some of the yarn.lock changes, I think they're unrelated? They're all about MDX stuff

LekoArts commented:
> Unless these test failures were caused by some of the yarn.lock changes, I think they're unrelated

You (or probably just yarn itself) changed a bunch of babel dependencies and the MDX dependency. @wardpeet didn't you hit this issue too last week or so?

me4502 commented:
> This would get added to every Gatsby site. It seems like a good idea but isbot is 3kb or so https://bundlephobia.com/package/isbot@3.4.5
> 
> Perhaps there's a lighter weight way of doing this?

Yeah if we just want to drop GoogleBot and the other big search engines we can just put in a regex to purely match those. Tbh given the amount of legitimately useful things that isbot would block this for, I'm heavily leaning towards this anyway. I'll make that change.

me4502 commented:
I've moved the call site to a better place, made it just check a simple regex, and made it relatively defensive. I also added a test to ensure that this actually does prevent prefetching.

The tests that are currently failing, such as the compilation-hash one, do confuse me. From all of my testing, there is no runtime change to how they function with my changes, as they never trigger the false return here. Is there something deep in Gatsby that I'm missing that might cause these tests to fail in these cases?

me4502 commented:
I've merged master into it and some tests have failed again.

I'm not convinced these are related, seeing as they sometimes pass (they were passing prior to merging master). When running locally what the tests are doing work fine, just not always when run within the test runner. On top of that, these tests also sometimes fail for me on master locally.

These tests are also about the build process (one of which seems to be that the logging format changed?), and my change is purely touching client-side files

wardpeet commented on packages/gatsby/cache-dir/loader.js:
```diff
@@ -388,6 +391,16 @@ export class BaseLoader {
       return false
     }
 
+    // Don't prefetch if this is a crawler bot
+    if (
+      `userAgent` in navigator &&
+      typeof navigator.userAgent !== `undefined`
+    ) {
+      if (BOT_REGEX.test(navigator.userAgent)) {
+        return false
+      }
+    }
```
This is a good enough check and saves a few bytes.
```suggestion
    // Don't prefetch if this is a crawler bot
    if (
     navigator.userAgent && BOT_REGEX.test(navigator.userAgent)
    ) {
      return false
    }
```

LekoArts commented:
> I've merged master into it and some tests have failed again.

They were flaky, so I restarted them :)
